packages/shared/plugins/PerfOperations/utils.ts

```diff
@@ -0,0 +1,144 @@
+import {
+  $getNodeByKey,
+  $isElementNode,
+  EditorState,
+  ElementNode,
+  LexicalNode,
+  SerializedElementNode,
+  SerializedLexicalNode,
+} from "lexical";
+import { PerfKind } from "./types";
+import { SerializedUsfmElementNode, UsfmElementNode } from "../../nodes/UsfmElementNode";
+/**
+ * this module provides utility functions for the Perf plugin.
+ * @module
+ */
+
+/**
+ * Finds the closest non-orphan ancestor node and returns its key.
+ * @param nodeKey - The key of the node to check.
+ * @param editorState - The current editor state.
+ * @returns The key of the closest non-orphan ancestor node, or null if no such ancestor exists.
+ * @remarks This function drills up the tree to find the closest non-orphan ancestor node. if the current node has a parent, it returns the current node's key. If the current node does not have a parent, it recursively calls itself with the parent node's key until it finds a non-orphan ancestor and returns its key.
+ */
+export function findClosestNonOrphanAncestorKey(
+  nodeKey: string,
+  editorState: EditorState,
+): string | null {
+  const node = editorState._nodeMap.get(nodeKey);
+  if (!node) return null;
+  const parentKey = node.__parent;
+  if (!parentKey) return nodeKey;
+  const hadParent = !!$getNodeByKey(parentKey); // Did its current parent exist
+  return hadParent ? nodeKey : findClosestNonOrphanAncestorKey(parentKey, editorState);
+}
+
+/**
+ * Recursively gets the perf path from the given lexical node .
+ * @param node - The node to get the path for.
+ * @param pathArray - The array representing the path to the node.
+ * @returns The array representing the path to the node.
+ */
+export function recursiveGetNodePath(
+  node: UsfmElementNode,
+  pathArray: Array<string | number> = [],
+): Array<string | number> {
+  const parent: UsfmElementNode | null = node.getParent();
+  const kind = getPerfKindFromNode(node);
+  const nodeKey = node.getKey();
+  const index = parent ? parent.getChildrenKeys?.().findIndex((key) => nodeKey === key) : 0;
+
+  if (kind === PerfKind.Sequence) {
+    const target = node.getAttributes?.()?.["data-target"];
+    pathArray.unshift(target ?? "main"); // Add the target to the path array
+    return pathArray;
+  }
+
+  pathArray.unshift(index); // Add the index to the path array
+
+  if (kind === PerfKind.ContentElement) {
+    pathArray.unshift("content"); // Add "content" to the path array
+  }
+
+  if (kind === PerfKind.Block) {
+    pathArray.unshift("blocks"); // Add "blocks" to the path array
+  }
+
+  return parent ? recursiveGetNodePath(parent, pathArray) : pathArray; // Recursively call the function with the parent node
+}
+
+// Function to determine the perf kind based on the path length
+export function getPerfKindbyPathLength(len: number): PerfKind {
+  switch (true) {
```

Comment:
hahaha you are right, this is because I used to have a condition that was using greater than (>) so I read somewhere about a way to use switch-case like this by switching on true, and tried it. Anyway, this function will be removed too, there is a new function to get the path, this was orphaned.